sputnikdao2/src/upgrade.rs

```diff
@@ -118,51 +115,24 @@ pub(crate) fn upgrade_using_factory(code_hash: Base58CryptoHash) {
 }
 
 #[allow(dead_code)]
-/// Self upgrade, optimizes gas by not loading into memory the code.
 pub(crate) fn upgrade_self(hash: &[u8]) {
     let current_id = env::current_account_id();
-    let method_name = "migrate".as_bytes().to_vec();
     let attached_gas = env::prepaid_gas() - env::used_gas() - GAS_FOR_UPGRADE_SELF_DEPLOY;
-    unsafe {
-        // Load input (wasm code) into register 0.
-        sys::storage_read(hash.len() as _, hash.as_ptr() as _, 0);
-        // schedule a Promise tx to this same contract
-        let promise_id = sys::promise_batch_create(
-            current_id.as_bytes().len() as _,
-            current_id.as_bytes().as_ptr() as _,
-        );
-        // 1st item in the Tx: "deploy contract" (code is taken from register 0)
-        sys::promise_batch_action_deploy_contract(promise_id, u64::MAX as _, 0);
-        // 2nd item in the Tx: call this_contract.migrate() with remaining gas
-        sys::promise_batch_action_function_call(
-            promise_id,
-            method_name.len() as _,
-            method_name.as_ptr() as _,
-            0 as _,
-            0 as _,
-            0 as _,
-            attached_gas.0,
-        );
-    }
+    let input = env::storage_read(hash).expect("ERR_NO_HASH");
+    let promise_id = env::promise_batch_create(&current_id);
+    env::promise_batch_action_deploy_contract(promise_id, &input);
+    env::promise_batch_action_function_call(promise_id, "migrate", &[], NO_DEPOSIT, attached_gas);
 }
 
 pub(crate) fn upgrade_remote(receiver_id: &AccountId, method_name: &str, hash: &[u8]) {
-    unsafe {
-        // Load input into register 0.
-        sys::storage_read(hash.len() as _, hash.as_ptr() as _, 0);
-        let promise_id = sys::promise_batch_create(
-            receiver_id.as_bytes().len() as _,
-            receiver_id.as_bytes().as_ptr() as _,
-        );
-        let attached_gas = env::prepaid_gas() - env::used_gas() - GAS_FOR_UPGRADE_REMOTE_DEPLOY;
-        sys::promise_batch_action_function_call(
-            promise_id,
-            method_name.len() as _,
-            method_name.as_ptr() as _,
-            u64::MAX as _,
-            0 as _,
-            0 as _,
-            attached_gas.0,
-        );
-    }
+    let input = env::storage_read(hash).expect("ERR_NO_HASH");
+    let promise_id = env::promise_batch_create(receiver_id);
+    let attached_gas = env::prepaid_gas() - env::used_gas() - GAS_FOR_UPGRADE_REMOTE_DEPLOY;
+    env::promise_batch_action_function_call(
+        promise_id,
+        method_name,
+        &input,
+        NO_DEPOSIT,
```

Comment:
@ctindogaru i get that we're transitioning this, however - shouldn't this accept attached deposit? certain methods this calls require deposits

Reply:
Basically, it should just fwd any/all deposits